### PR TITLE
Mistake in parameters' grad norm tracking

### PR DESCRIPTION
**Pull request by ivannz**
There is a mistake in [grad_norms](https://github.com/PyTorchLightning/pytorch-lightning/blob/fdbbe968256f6c68a5dbb840a2004b77a618ef61/pytorch_lightning/core/grads.py#L11) in `core.grads.GradInformation`. The mistake affects the reported gradient norm for every individual parameters, but not the total norm. This PR fixes the erroneous computation.

According to torch [docs](https://pytorch.org/docs/stable/torch.html#torch.norm) `torch.norm(tensor, p)` computes the vector norm: either 2-norm, or `sum(abs(x)**p)**(1./p)` if p != 2, and flattens the tensor if required. Now `p.grad.data.norm` in [grads.py#L17](https://github.com/PyTorchLightning/pytorch-lightning/blob/fdbbe968256f6c68a5dbb840a2004b77a618ef61/pytorch_lightning/core/grads.py#L17) is the same as `torch.norm(p.grad.data, ...)`, and thus computes the norm. However on [grads.py#L19](https://github.com/PyTorchLightning/pytorch-lightning/blob/fdbbe968256f6c68a5dbb840a2004b77a618ef61/pytorch_lightning/core/grads.py#L19) grad tracker takes the `p-th` root again, essentially making `params_norm` equal to `sum(abs(x)**p)**(1./(p**2)) = norm**(1./p)` which is not correct.

The following snippet, which borrows code from grads.py#L17-L19
```python
import torch
import numpy as np

x = torch.randn(3, 4, 32, 32) * 0.1
for norm_type in [1., 1.5, 2., 4.]:
    # numpy reference value
    np_val = np.linalg.norm(x.numpy().ravel(), norm_type)

    # according to docs (and definition of p-norm)
    tr_val_naive = (abs(x) ** norm_type).sum().pow(1./norm_type)

    # almost verbatim from grads.py#L17-L19
    param_norm = x.norm(norm_type)
    grad_norm = param_norm ** (1./norm_type)

    print(f'numpy.norm = {float(np_val):4g}, '
          f'torch.naive = {float(tr_val_naive):4g}, '
          f'grad_norm = {float(grad_norm):4g}')
```
produces 
```
numpy.norm = 972.305, torch.naive = 972.305, grad_norm = 972.306
numpy.norm = 47.8891, torch.naive = 47.8891, grad_norm = 13.1874
numpy.norm = 11.0455, torch.naive = 11.0455, grad_norm = 3.32348
numpy.norm = 1.38376, torch.naive = 1.38376, grad_norm = 1.08459
```
which is clearly incorrect.

### CODE REVIEW

**Comment by ivannz** [comment]
If necessary there is a patch to this fix that would enable support of infinity-norm computation:
1. replace `total_norm` with a list and accumulate individual norms in it
2. use `total_norm = torch.tensor(total_norm).norm(norm_type)`

**Comment by williamFalcon** [comment]
@ivannz excellent! mind adding a test to make sure we're matching what numpy expects?

**Comment by ivannz** [comment]
I can add a unit test, but I have a question. Since the latest code relies completely on `torch.norm`, it is it necessary to test if `torch.norm` coincides with `np.linalg.norm`? It would seem that pytorch's ATen should already test its implementation against BLAS.

**Comment by williamFalcon** [comment]
yes, but we need to make sure the overall result of this gradient clipping and the norm calculation are correct... we've addressed this issue multiple times already and at this point we just need more rigorous testing on this so that we don't have to revisit...

**Comment by ivannz** [comment]
I have added a test, @williamFalcon .

Due to explicit rounding to 3 places in `grad_norm` the test uses `5e-3` relative tolerance for checking the norms.

File: pytorch_lightning/core/grads.py [hunk 1]
```diff
@@ -3,28 +3,25 @@
...
 from typing import Dict
 
-from torch import nn
+import torch
 
 
-class GradInformation(nn.Module):
+class GradInformation(torch.nn.Module):
 
     def grad_norm(self, norm_type: float) -> Dict[str, int]:
```
Comment (awaelchli) [inline pytorch_lightning/core/grads.py:11]: Would you be so kind to also add a brief docstring describing which values are allowed for "norm type" (p >= 1) and will be returned in the dict?

File: tests/models/test_grad_norm.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+from pytorch_lightning import Trainer
+
+from pytorch_lightning.loggers import LightningLoggerBase
+from pytorch_lightning.utilities import rank_zero_only
+
+from tests.base import EvalModelTemplate
+
+
+class OnlyMetricsListLogger(LightningLoggerBase):
```
Comment (awaelchli) [inline tests/models/test_grad_norm.py:13]: do we really need this? why is the default logger not applicable here?

File: tests/models/test_grad_norm.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+
+class ModelWithManualGradTracker(EvalModelTemplate):
+    def __init__(self, norm_type, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.stored_grad_norms, self.norm_type = [], norm_type
+
+    # validation spoils logger's metrics with `val_loss` records
+    validation_step = None
+    val_dataloader = None
```
Comment (awaelchli) [inline tests/models/test_grad_norm.py:50]: I see, I think you want the two dicts to be exactly the same, but I think the test is valid if we test that stored_grad_norms is a subset of the one logged. What do you think? This would simplify the test without compromising the validity of it.

File: tests/models/test_grad_norm.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+            out[prefix + name] = round(norm, 3)
+
+        # handle total norm
+        norm = np.linalg.norm(norms, self.norm_type)
+        out[prefix + 'total'] = round(norm, 3)
+        self.stored_grad_norms.append(out)
+
+
+# @pytest.mark.skip(reason="might fail for small `norm_type` due to round-off")
```
Comment (awaelchli) [inline tests/models/test_grad_norm.py:78]: I would set the seed (see other tests how it is done) to a value that works so that the test does not randomly crash if the weights initialization give a too small norm. Makes sense?

File: pytorch_lightning/core/grads.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/core/grads.py:11]: ... and also add a changelog entry for this bugfix?

Comment (awaelchli) [reply pytorch_lightning/core/grads.py:11]: The annotation should also be changd to Dict[str, float]

File: tests/models/test_grad_norm.py (hunk 3 above)
Comment (ivannz) [reply tests/models/test_grad_norm.py:50]: Not exactly, since I use lists to store norm-dicts.

Logger's `log_metrics` stores whatever the caller requests in a list. I have found relevant two points where `log_metrics` is invoked: in [trainer.run_training_epoch()](https://github.com/PyTorchLightning/pytorch-lightning/blob/0be530a42774a9a7868ab1f7ba3f6392d6fcde9e/pytorch_lightning/trainer/training_loop.py#L483) and in [trainer.run_evaluation()](https://github.com/PyTorchLightning/pytorch-lightning/blob/0be530a42774a9a7868ab1f7ba3f6392d6fcde9e/pytorch_lightning/trainer/evaluation_loop.py#L385). In lockstep with training loop the custom model updates its own list of grad norm dicts in `on_after_backward` called by [trainer.run_training_batch()](https://github.com/PyTorchLightning/pytorch-lightning/blob/0be530a42774a9a7868ab1f7ba3f6392d6fcde9e/pytorch_lightning/trainer/training_loop.py#L609) from [trainer.run_training_epoch()](https://github.com/PyTorchLightning/pytorch-lightning/blob/0be530a42774a9a7868ab1f7ba3f6392d6fcde9e/pytorch_lightning/trainer/training_loop.py#L438)). Hence, in order to keep the logger's list synced with model's list I decided to disable validation. IMO, this is a cleaner solution than manually *deinterleaving* validation dicts and grad norm dicts from logger's list.

File: tests/models/test_grad_norm.py (hunk 2 above)
Comment (ivannz) [reply tests/models/test_grad_norm.py:13]: I have inspected the [LightningLoggerBase](https://github.com/PyTorchLightning/pytorch-lightning/blob/0be530a42774a9a7868ab1f7ba3f6392d6fcde9e/pytorch_lightning/loggers/base.py#L132), and it explicitly declares `log_metrics` abstract. the [DummyLogger](https://github.com/PyTorchLightning/pytorch-lightning/blob/0be530a42774a9a7868ab1f7ba3f6392d6fcde9e/pytorch_lightning/loggers/base.py#L304) completely ignores the data passed to `log_metrics` acting essentially like `... > /dev/null`. I couldn't find any other class derived from `LightningLoggerBase` that would keep list of the data records passed to it, except for the [CustomLogger](https://github.com/PyTorchLightning/pytorch-lightning/blob/0be530a42774a9a7868ab1f7ba3f6392d6fcde9e/tests/loggers/test_base.py#L30) which unfortunately only keeps the most recent records (and is somewhat outside of the package). Thus I simply implemented a special logger for the sole purpose of keeping history of the dicts passed to `log_metrics`.

File: tests/models/test_grad_norm.py (hunk 4 above)
Comment (ivannz) [reply tests/models/test_grad_norm.py]: done

**Comment by ivannz** [comment]
To reflect similarity in spirit with [clip_grad_norm](https://github.com/pytorch/pytorch/blob/54d4b419db8962a01978a6c191148f0174080978/torch/nn/utils/clip_grad.py#L6) I added support for an explicit 'inf' string in trainer's `track_grad_norm` argument: it can be an `int`, `float` or an `'inf'` for an infinity-norm (or negative to disable norm tracking). The value is type-cheked in `Trainer.__init__` (with `MisconfigurationException`) and cast to float on every check in [Trainer.run_training_batch()](https://github.com/PyTorchLightning/pytorch-lightning/blob/0be530a42774a9a7868ab1f7ba3f6392d6fcde9e/pytorch_lightning/trainer/training_loop.py#L628).

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -341,7 +341,13 @@ def __init__(
             self.gradient_clip = gradient_clip
 
         self.check_val_every_n_epoch = check_val_every_n_epoch
-        self.track_grad_norm = track_grad_norm
+
+        if not isinstance(track_grad_norm, (int, float)) \
+           and track_grad_norm != 'inf':
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:346]:
```suggestion
        if not isinstance(track_grad_norm, (int, float)) and track_grad_norm != 'inf':
```

File: pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -341,7 +341,13 @@ def __init__(
...
+            raise MisconfigurationException("track_grad_norm can be an int, a "
+                                            "float or 'inf' (infinity norm).")
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:348]:
```suggestion
            raise MisconfigurationException(
                "track_grad_norm can be an int, a float or 'inf' (infinity norm).")
```

File: tests/models/test_grad_norm.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    # use a custom grad tracking module and a list logger
+    model = ModelWithManualGradTracker(norm_type)
+    logger = OnlyMetricsListLogger()
+
+    result = Trainer(
+        max_epochs=3,
+        logger=logger,
+        track_grad_norm=norm_type,
+        row_log_interval=1,  # request grad_norms every batch
+    ).fit(model)
```
Comment (Borda) [inline tests/models/test_grad_norm.py:94]:
```suggestion
    trainer = Trainer(
        max_epochs=3,
        logger=logger,
        track_grad_norm=norm_type,
        row_log_interval=1,  # request grad_norms every batch
    )
    result = trainer.fit(model)
```

File: tests/models/test_grad_norm.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        out[prefix + 'total'] = round(norm, 3)
+        self.stored_grad_norms.append(out)
+
+
+@pytest.mark.parametrize("seed", [479_158_593])  # a vetted random number
+@pytest.mark.parametrize("norm_type", [1., 1.25, 1.5, 2, 3, 5, 10, 'inf'])
+def test_grad_tracking(tmpdir, norm_type, seed, rtol=5e-3):
+    # rtol=5e-3 respects the 3 decmials rounding in `.grad_norms` and above
+
+    seed_everything(seed)
```
Comment (Borda) [inline tests/models/test_grad_norm.py:83]:
```suggestion
@pytest.mark.parametrize("norm_type", [1., 1.25, 1.5, 2, 3, 5, 10, 'inf'])
def test_grad_tracking(tmpdir, norm_type, rtol=5e-3):
    # rtol=5e-3 respects the 3 decmials rounding in `.grad_norms` and above

    reset_seed()
```

File: tests/models/test_grad_norm.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
+import torch
+import pytest
+import numpy as np
+
+from pytorch_lightning import Trainer, seed_everything
+
+from pytorch_lightning.loggers import LightningLoggerBase
+from pytorch_lightning.utilities import rank_zero_only
+
+from tests.base import EvalModelTemplate
```
Comment (Borda) [inline tests/models/test_grad_norm.py:10]:
```suggestion
from tests.base import EvalModelTemplate
from tests.base.utils import reset_seed
```

**Comment by ivannz** [comment]
@awaelchli and @Borda thank you for the code review!

@Borda Thank you for your edits! As for your doc string style fix, I apologise for not having read `CONTRBUTING.md` carefully.

**Comment by Borda** [comment]
> @Borda Thank you for your edits! As for your doc string style fix, I apologise for not having read `CONTRBUTING.md` carefully.

that is fine, we thank you for your contribution :]
